code/week6/findBlue.cpp

```diff
@@ -0,0 +1,81 @@
+#include <unistd.h>
+#include <opencv2/opencv.hpp>
+#include <opencv2/core/core.hpp>
+#include <opencv2/highgui/highgui.hpp>
+#include <opencv2/highgui/highgui.hpp>
+
+using namespace cv;
+
+
+Mat findBlue(const Mat& frameBGR) {
+      //cv::Scalar is a container that is used to hold pixel values
+      //the values shown below are tuned by hand on example data that is
+      //gathered
+      const Scalar blue_low{78, 50, 70};     //these are the lowest HSV values
+                                             //that we accept as blue
+      const Scalar blue_high{138, 255, 255}; //these are the highest HSV
+                                             //values we accept as blue
+
+      // kernel that is an ellipse of size 11X11 pixels
+      Mat erosion_kernel_blue = getStructuringElement(MORPH_ELLIPSE, Size(11, 11));
+
+      //applying a GaussianBlur makes the image "fuzzier". This smoothes out
+      //noisy pixels
+      Mat frameBlurred;
+      GaussianBlur(frameBGR, frameBlurred, Size{7,7}, 0);
+
+      //We shift the mat which is encodes as an RGB image into an HSV image
+      Mat frameHSV;
+      cvtColor(frameBlurred, frameHSV, CV_BGR2HSV);
+
+      //We initialize an empty Mat to store all the pixels that we identify
+      //as blue. CV_8U is the image depth: each pixel is only one bit
+      Mat output_blue = Mat::zeros(200, 640, CV_8U);
+
+      //for every pixel in our frame that we define as blue, write a white
+      //pixel to the output_blue Mat
+      inRange(frameHSV, blue_low, blue_high, output_blue);
+
+      //apply an erode operation onto the filtered image. This helps us
+      //remove specks of blue that show up due to noise or glare
+      erode(output_blue, output_blue, erosion_kernel_blue);
+
+      //the logic to publish the result of the filtering in ROS would go here
+    return output_blue;
+}
+
+int main() {
+    // sets up the input stream to the default video device
+    // /dev/video0 must exist
+    VideoCapture stream1(0);
+
+    // this checks to see if a key has been pressed and then shows the image for
+    // one millisecond
+    while (waitKey(1) == 255) {
```

Comment:
For me this causes the image to never come up at all. You probably want:

`waitKey(1) == -1`.

I also take back this being an OK solution, since you might want to press a key and have the program keep running (for example, my window dragging needs to have a key pressed, so I'm unable to manipulate the window).

Perhaps, you could use:
```
(char)waitKey(1) != 'q'
```
or similar.